File: VB/BindingModeToPropertyTemplateConverter.vb
Imports DevExpress.XtraReports.UI
Imports System
Imports System.Globalization
Imports System.Windows
Imports System.Windows.Data

Namespace SmartTagCustomization

    Public Class BindingModeToPropertyTemplateConverter
        Implements IValueConverter

        Public Property BindingTemplate As DataTemplate

        Public Property ExpressionTemplate As DataTemplate

        Public Function Convert(ByVal value As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IValueConverter.Convert
            Dim mode = CType(value, DataBindingMode?)
            If Not mode.HasValue Then Return Nothing
            If mode.Value = DataBindingMode.Bindings Then
                Return BindingTemplate
            Else
                Return ExpressionTemplate
            End If
        End Function

        Private Function ConvertBack(ByVal value As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IValueConverter.ConvertBack
            Throw New NotSupportedException()
        End Function
    End Class
End Namespace

File: CS/BindingModeToPropertyTemplateConverter.cs
using DevExpress.XtraReports.UI;
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Text;
using System.Windows;
using System.Windows.Data;

namespace SmartTagCustomization {
    public class BindingModeToPropertyTemplateConverter : IValueConverter {
        public DataTemplate BindingTemplate { get; set; }
        public DataTemplate ExpressionTemplate { get; set; }
        public object Convert(object value, Type targetType, object parameter, CultureInfo culture) {
            var mode = value as DataBindingMode?;
            if (!mode.HasValue)
                return null;
            if (mode.Value == DataBindingMode.Bindings)
                return BindingTemplate;
            else return ExpressionTemplate;
        }

        object IValueConverter.ConvertBack(object value, Type targetType, object parameter, CultureInfo culture) {
            throw new NotSupportedException();
        }
    }
}
